### PR TITLE
fix a corner case of torch.arange

### PR DESCRIPTION
**Pull request by yuguo68**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80758

trying to fix https://github.com/pytorch/pytorch/issues/80589, see the two corner cases in the issue. 
added the two cases in unit tests and add device to the tests. 

Edit: the landed version only fixes the first issue in https://github.com/pytorch/pytorch/issues/80589

### CODE REVIEW

**Comment by ezyang** [comment]
It would have been nice for the device parametrization to be separately stacked (no need to do now, I'll review as is)

File: test/test_tensor_creation_ops.py [hunk 1]
```diff
@@ -2472,124 +2472,143 @@ def test_range_warning(self, device):
...
+        res2 = torch.tensor([], device=device)
         torch.arange(1, 0, -1, out=res2)
         self.assertEqual(res1, res2, atol=0, rtol=0)
         torch.arange(1, 2, 1, out=res2)
         self.assertEqual(res1, res2, atol=0, rtol=0)
 
         # FloatTensor
-        res1 = torch.arange(0.6, 0.89, 0.1, out=torch.FloatTensor())
+        out = torch.tensor([], dtype=torch.float, device=device)
+        res1 = torch.arange(0.6, 0.89, 0.1, out=out)
```
Comment (ezyang) [inline test/test_tensor_creation_ops.py:2516]: good riddance, thx

File: aten/src/ATen/native/RangeFactories.cpp [hunk 2]
```diff
@@ -143,12 +143,20 @@ Tensor& range_out(const Scalar& start, const Scalar& end, const Scalar& step, Te
...
 
 Tensor& arange_out(const Scalar& start, const Scalar& end, const Scalar& step, Tensor& result) {
+  bool args_all_int = start.isIntegral(true) && end.isIntegral(true) && step.isIntegral(true);
   AT_DISPATCH_ALL_TYPES_AND2(kHalf, kBFloat16, result.scalar_type(), "arange_cpu", [&]() {
     using accscalar_t = at::acc_type<scalar_t, false>;
     auto xstart = start.to<accscalar_t>();
     auto xend = end.to<accscalar_t>();
     auto xstep = step.to<accscalar_t>();
 
+    TORCH_CHECK(xstep > 0 || xstep < 0, "step must be nonzero");
```
Comment (ezyang) [inline aten/src/ATen/native/RangeFactories.cpp:153]: Lol are there some floating point shenanigans why `xstep != 0` is invalid here

Comment (ezyang) [reply aten/src/ATen/native/RangeFactories.cpp:153]: preexisting condition

**Review by ezyang — APPROVED** [review]

File: test/test_tensor_creation_ops.py [hunk 3]
```diff
@@ -2472,124 +2472,143 @@ def test_range_warning(self, device):
...
+        r = torch.arange(0, 5, device=device)
         self.assertEqual(r.min(), 0)
         self.assertEqual(r.max(), 4)
         self.assertEqual(r.numel(), 5)
 
-        r = torch.arange(0, 5, 2)
+        r = torch.arange(0, 6, 3, device=device)
+        self.assertEqual(r.min(), 0)
+        self.assertEqual(r.max(), 3)
+        self.assertEqual(r.numel(), 2)
```
Comment (ezyang) [inline test/test_tensor_creation_ops.py:2551]: this is the new test

File: test/test_tensor_creation_ops.py [hunk 4]
```diff
@@ -2472,124 +2472,143 @@ def test_range_warning(self, device):
...
+        r = torch.arange(0, 5, 2, device=device)
         self.assertEqual(r.min(), 0)
         self.assertEqual(r.max(), 4)
         self.assertEqual(r.numel(), 3)
 
-        r1 = torch.arange(0, 5 + 1e-6)
+        r = torch.arange(0, -5, -2, device=device)
+        self.assertEqual(r.min(), -4)
+        self.assertEqual(r.max(), 0)
+        self.assertEqual(r.numel(), 3)
```
Comment (ezyang) [inline test/test_tensor_creation_ops.py:2561]: and this too, the test here is so bad woof

**Comment by ezyang** [comment]
onnx failure looks real

**Comment by yuguo68** [comment]
>

**Comment by yuguo68** [comment]
> onnx failure looks real

thanks for reviewing. Yes, it turns out in onnx we always cast the args (start, end, step) to desired dtype before computing size. 
On master, 
`
>>> import torch
>>> import numpy as np
>>> np.arange(1, 5.5, 1.5, dtype=np.int32)
array([1, 2, 3], dtype=int32)
>>> np.arange(1, 5.5, 1.5, dtype=np.int64)
array([1, 2, 3])
>>> torch.arange(1, 5.5, 1.5, dtype=torch.int32)
tensor([1, 2, 3], dtype=torch.int32)
>>> torch.arange(1, 5.5, 1.5, dtype=torch.int64)
tensor([1, 2, 3, 4])
`
This PR will change the last one to 
`
>>> torch.arange(1, 5.5, 1.5, dtype=torch.int64)
tensor([1, 2, 3])
`
but for onnx, 
'
>>> onnx.arange(1, 5.5, 1.5, dtype=torch.int32)
tensor([1, 2, 3, 4], dtype=torch.int32)
>>> onnx.arange(1, 5.5, 1.5, dtype=torch.int64)
tensor([1, 2, 3, 4])
`
the failed unittest compare torch and onnx with int64, so it works on master but fails with this PR. We may need to change the implementation in torch.onnx to make it consistent.

**Pull request closed without merging**